src/ifcsverchok/nodes/ifc/add_spatial_element.py

```diff
@@ -0,0 +1,149 @@
+import bpy
+
+import ifcopenshell
+import ifcsverchok.helper
+from ifcsverchok.ifcstore import SvIfcStore
+
+from bpy.props import StringProperty
+from sverchok.node_tree import SverchCustomTreeNode
+from sverchok.data_structure import updateNode, flatten_data, repeat_last_for_length, ensure_min_nesting
+
+
+class SvIfcAddSpatialElement(bpy.types.Node, SverchCustomTreeNode, ifcsverchok.helper.SvIfcCore):
+    bl_idname = "SvIfcAddSpatialElement"
+    bl_label = "IFC Add Spatial Element"
+    node_dict = {}
+ 
+    Names: StringProperty(name="Name(s)", update=updateNode)
+    IfcClass: StringProperty(name="IFC Class", update=updateNode, default="IfcSpace")
+    Elements: StringProperty(name="Elements", update=updateNode)
+
+    def sv_init(self, context):
+        self.inputs.new("SvStringsSocket", "Names").prop_name = "Names"
+        self.inputs.new("SvStringsSocket", "IfcClass").prop_name = "IfcClass"
+        self.inputs.new("SvStringsSocket", "Elements").prop_name = "Elements"
+        self.outputs.new("SvStringsSocket", "Entities")
+        self.node_dict[hash(self)] = {}
+    
+    def draw_buttons(self, context, layout):
+        layout.operator("node.sv_ifc_tooltip", text="", icon="QUESTION", emboss=False).tooltip = "Ifc entity by type."
+
+    def process(self):
+        self.sv_input_names = [i.name for i in self.inputs]
+        if hash(self) not in self.node_dict:
```

Comment:
Why `hash(self)` and not just `self`? This generates collisions. dicts are are hash maps, so you don't need to manually hash the key. Internally in a hash map the hash is used to determine the "bucket" the value will go into, but there will still be actual equality testing within the bucket. As the hash is not necessarily unique.

Reply:
In short, hash(self) was the only number/ID that I could find was, and stayed, unique for each node instance. Sverchok uses the `node_id` attribute internally, which is also based on hash(self) (https://github.com/nortikin/sverchok/issues/2445). And I don't remember exactly why right now, but I had some issues with that attribute either not staying unique or changing during the life time of the node, but didn't have the same issues with using hash(self) directly. I would have to test it again.
I'm not sure how you would do it with just `self`? The issue with not having an internal node_dict was, that two nodes of the same type (Eg. two "IFC By Type" nodes) will not make new class instances, but both use the same one, so stored information would leak from one node to another. If you have an idea on how this can be handled differently, I'd like to change it. Not sure if the problem mentioned [here](https://github.com/nortikin/sverchok/issues/3564#issuecomment-695770526) also applies?

Reply:
That approach there would indeed not stay unique as it explicitly `xor`s in the current time. I'm also not sure how authoritative that post is. You can see explained here what the relationship is between id() [what internally would be used when you use `self`] and hash() for types that don't override __hash__ https://stackoverflow.com/a/33572401 Python may recycle ids when they are disposed so that's why they factor in the current time. But the problem remains. I still think `self` here is better than `hash(self)`. If you want a numeric ID, the proper way would probably be a globally defined auto incrementing counter https://stackoverflow.com/a/1045724

Anyway this is a bit of a rabbit hole. I'm fine if we keep it like this. Just wanted to provide some bg info. I'll mark this as resolved.